lib/rosette/integrations/smartling-integration/translation_memory_builder.rb

```diff
@@ -47,76 +38,42 @@ def set_logger(logger)
         end
 
         def build
-          phrases = get_phrases
-          @uploader = build_uploader_for(phrases)
-
-          logger.info('Uploading translation memory seed to Smartling')
-          uploader.upload
-
           TranslationMemory.new(
-            download(repo_config.locales, uploader)
+            if thread_pool_size > 0
+              build_asynchronously
+            else
+              build_synchronously
+            end
           )
         end
 
-        # used?
-        def cleanup
-          if uploader
-            delete_file(uploader.destination_file_uri)
-          end
-        end
-
         private
 
-        def hash_mutex
-          @hash_mutex ||= Mutex.new
-        end
-
-        def get_phrases
-          if thread_pool_size > 0
-            get_phrases_asynchronously
-          else
-            get_phrases_synchronously
-          end
-        end
-
-        def get_phrases_synchronously
-          datastore.each_unique_meta_key(repo_config.name).each_with_object({}) do |meta_key, ret|
-            recent_key = datastore.most_recent_key_for_meta_key(
-              repo_config.name, meta_key
-            )
-
-            ret[meta_key] = recent_key
+        def build_synchronously
+          repo_config.locales.each_with_object({}) do |locale, ret|
+            ret[locale.code] = download_and_process(locale)
           end
         end
 
-        def get_phrases_asynchronously
+        def build_asynchronously
           pool = Concurrent::FixedThreadPool.new(thread_pool_size)
-          counter = 0
+          hash_mutex ||= Mutex.new
+          total = 0
 
-          phrases = datastore.each_unique_meta_key(repo_config.name).each_with_object({}) do |meta_key, ret|
+          result = repo_config.locales.each_with_object({}) do |locale, ret|
             pool << Proc.new do
-              recent_key = datastore.most_recent_key_for_meta_key(
-                repo_config.name, meta_key
-              )
-
-              hash_mutex.synchronize do
-                ret[meta_key] = recent_key
-              end
+              memory_hash = download_and_process(locale)
```

Comment:
Unfortunately not, since `build_synchronously` and `build_asynchronously` both loop through locales. In the case of `build_asynchronously`, a bunch of `Proc`s are added to a thread pool, which `build_synchronously` doesn't do. The `download_and_process` method handles building for a single locale, which both methods call.